### PR TITLE
🍤 WhatsApp Templates

### PR DESCRIPTION
**Pull request by nicpottier**
Adds `Template` and `ChannelTemplate` models. Right now only WhatsApp syncs these and creates them based on the Facebook account data. Also adds API endpoint to get the current set of templates for an org.

### CODE REVIEW

File: temba/channels/types/whatsapp/tasks.py [hunk 1]
```diff
@@ -84,3 +85,86 @@ def refresh_whatsapp_tokens():
...
+                        content=template["content"],
+                        variable_count=_calculate_variable_count(template["content"]),
+                        status=STATUS_MAPPING[template["status"]],
+                        external_id=template["id"],
+                    )
+
+                    seen.append(template)
+
+                # trim any templates we didn't see
+                ChannelTemplate.trim(channel, seen)
```
Comment (ericnewcomer) [inline temba/channels/types/whatsapp/tasks.py]: How do we want to deal with flows that reference templates that are trimmed? Should we create a dependency and clean up flow definitions when this happens?

Comment (nicpottier) [reply temba/channels/types/whatsapp/tasks.py]: I'm not sure what we can do automatically. Deletion of a channel template happens at the Facebook side and not like we want to remove the flow entirely when that happens.

There are two models here, `Template` which is the named part (and what will be referenced in the flow) and `ChannelTemplate` which is the actual translation per language (and has a FK to `Template`)) When we trim we are only trimming `ChannelTemplate` not `Template` as well.

So really what happens if they delete all the templates is that we won't be able to look up a valid template to substitute in and I think we back down to sending the normal text of the message (which may fail if the contact hasn't interacted in 24 hours).

Not sure what choice we really have on that front.

File: temba/api/v2/serializers.py [hunk 2]
```diff
@@ -1170,3 +1171,26 @@ def get_data(self, obj):
...
+                language=template.language,
+                content=template.content,
+                variable_count=template.variable_count,
+                status=template.get_status_display(),
+            )
+
+        return translations
+
+    class Meta:
+        model = WebHookEvent
```
Comment (rowanseymour) [inline temba/api/v2/serializers.py]: model = Template

File: temba/api/v2/serializers.py [hunk 3]
```diff
@@ -1170,3 +1171,26 @@ def get_data(self, obj):
...
+class TemplateReadSerializer(ReadSerializer):
+    name = serializers.CharField(required=True)
+    translations = serializers.SerializerMethodField()
+    modified_on = serializers.DateTimeField(default_timezone=pytz.UTC)
+    created_on = serializers.DateTimeField(default_timezone=pytz.UTC)
+
+    def get_translations(self, obj):
+        translations = dict()
+        for template in ChannelTemplate.objects.filter(template=obj).order_by("language"):
+            translations[template.language] = dict(
```
Comment (rowanseymour) [inline temba/api/v2/serializers.py]: I thought we agreed something about using `{"..": ""}` when returning JSON 😬

File: temba/api/v2/tests.py [hunk 4]
```diff
@@ -3620,3 +3621,42 @@ def test_flow_starts(self):
         # check filtering by id (deprecated)
         response = self.fetchJSON(url, "id=%d" % start2.id)
         self.assertResultsById(response, [start2])
+
+    def test_templates(self):
+        url = reverse("api.v2.templates")
+        self.assertEndpointAccess(url)
+
+        # create some templates
+        ChannelTemplate.ensure_exists(
```
Comment (rowanseymour) [inline temba/api/v2/tests.py]: is `ensure_exists` the new `get_or_create` ?

File: temba/templates/models.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    template = models.ForeignKey(Template, on_delete=models.PROTECT, related_name="channel_templates")
+
+    # the channel that synced this template
+    channel = models.ForeignKey(Channel, on_delete=models.PROTECT)
+
+    # the content of this template
+    content = models.CharField(max_length=1280, null=False)
+
+    # how many variables this template contains
+    variable_count = models.IntegerField()
```
Comment (rowanseymour) [inline temba/templates/models.py]: curious why this is a first-class thing since it's fairly easy to derive

File: temba/templates/models.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+from django.utils import timezone
+
+from temba.channels.models import Channel
+from temba.orgs.models import Org
+
+
+class Template(models.Model):
+    """
+    Templates represent messages that can be used in flows and have template variables substituted into them. These
+    are usually used by WhatsApp channels, but can also be used more generically to create DRY messages in flows.
```
Comment (rowanseymour) [inline temba/templates/models.py]: If we want these to be used generically, don't we need to de-couple the content from syncing over specific channels? Like have a `TemplateTranslation` model and then some other record of the state of syncing that to a particular channel?

File: temba/api/v2/serializers.py (hunk 3 above)
Comment (nicpottier) [reply temba/api/v2/serializers.py]: Oh sure, always forget that rule

File: temba/api/v2/tests.py (hunk 4 above)
Comment (nicpottier) [reply temba/api/v2/tests.py]: Happy to change that. It affects other models but probably clearer.

File: temba/templates/models.py (hunk 5 above)
Comment (nicpottier) [reply temba/templates/models.py]: Seemed a bit crazy to run regexes on everything in an API response.

File: temba/templates/models.py (hunk 6 above)
Comment (nicpottier) [reply temba/templates/models.py]: I mention WhatsApp here but you'll note everything whatsapp specific is done by the WhatsApp channel type.

File: temba/templates/models.py (hunk 5 above)
Comment (rowanseymour) [reply temba/templates/models.py]: yeah I'm just wondering who needs this? is it an editor thing?

File: temba/templates/models.py (hunk 6 above)
Comment (rowanseymour) [reply temba/templates/models.py]: yeah just trying to understand what it means that all template content is channel specific. I.e. you can't have a template with some translations that is syncable to multiple channels.

File: temba/templates/models.py (hunk 5 above)
Comment (nicpottier) [reply temba/templates/models.py]: Yes, editor needs it. Actually other reason to store is that it is specific to the template type. (IE, WhatsApp channel is the one that figures this out)

File: temba/templates/models.py (hunk 6 above)
Comment (nicpottier) [reply temba/templates/models.py]: Template content is channel specific but templates themselves aren't. So you could imagine two whatsapp channels with the same template name and have that work correctly. (though I find that unlikely) But ya, needs to be that way.

Comment (rowanseymour) [reply temba/templates/models.py]: (I forgot we don't create templates in RP, only pull them from channels)

File: templates/templates/template_list.haml [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+            %ul.nav.nav-list
+              %li.nav-header
+                -trans "Templates"
+
+              %li
+                %a{'href':'arst'}
+                  Synced
+              %li
+                %a{'href':'arst'}
+                  Synced
```
Comment (norkans7) [inline templates/templates/template_list.haml]: Synced repeated?

File: templates/templates/template_list.haml [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+          location.reload();
+        }
+      })
+      modal.show()
+    }
+    {% endif %}
+
+
+    {% if org_perms.templates.template_archive %}
+      $(".page-content").on('click', ".object-btn-archive", function() {
```
Comment (norkans7) [inline templates/templates/template_list.haml]: the class used above seems to be `template-btn-archive`

File: templates/templates/template_list.haml [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+	    runActionOnObjectRows("archive");
+      });
+    {% endif %}
+
+    {% if org_perms.templates.template_restore %}
+      $(".page-content").on('click', ".object-btn-restore", function() {
```
Comment (norkans7) [inline templates/templates/template_list.haml]: no button to restore

File: templates/templates/template_list.haml (hunk 8 above)
Comment (nicpottier) [reply templates/templates/template_list.haml]: Oops, these aren't even supposed to be in there (the templates), all the UI was removed. Thanks for catching it.

File: temba/api/v2/tests.py [hunk 10]
```diff
@@ -3620,3 +3621,50 @@ def test_flow_starts(self):
...
+
+    def test_templates(self):
+        url = reverse("api.v2.templates")
+        self.assertEndpointAccess(url)
+
+        # create some templates
+        TemplateTranslation.get_or_create(
+            self.channel, "hello", "eng", "Hi {{1}}", 1, TemplateTranslation.STATUS_APPROVED, "1234"
+        )
+        tt = TemplateTranslation.get_or_create(
```
Comment (rowanseymour) [inline temba/api/v2/tests.py]: good to add an object owned by another org in all these api unit tests and then check it doesn't appear in the results

File: temba/api/v2/views.py [hunk 11]
```diff
@@ -3116,3 +3121,86 @@ def get_write_explorer(cls):
             ],
             example=dict(body='{"flow":"f5901b62-ba76-4003-9c62-72fdacc1b7b7","urns":["twitter:sirmixalot"]}'),
         )
+
+
+class TemplatesEndpoint(ListAPIMixin, BaseAPIView):
+    """
+    This endpoint allows you to fetch the WhatsApp templates that have been synced. Each template contains a '
```
Comment (rowanseymour) [inline temba/api/v2/views.py]: '

File: temba/api/v2/views.py [hunk 12]
```diff
@@ -3116,3 +3121,86 @@ def get_write_explorer(cls):
...
+    pagination_class = ModifiedOnCursorPagination
+
+    def filter_queryset(self, queryset):
+        org = self.request.user.get_org()
+        template_ids = (
+            TemplateTranslation.objects.filter(channel__org=org, is_active=True)
+            .distinct("template_id")
+            .values_list("template_id", flat=True)
+        )
+        queryset = Template.objects.filter(org=org, id__in=template_ids).prefetch_related("translations")
```
Comment (rowanseymour) [inline temba/api/v2/views.py]: not sure why two queries are needed here.. are we trying to exclude templates without active translations? Maybe

```python
queryset = (
    org.templates.exclude(translations=None)
    .prefetch_related(Prefetch("translations", TemplateTranslation.objects.filter(is_active=True)))
)
```

File: temba/channels/types/whatsapp/tasks.py [hunk 13]
```diff
@@ -84,3 +85,83 @@ def refresh_whatsapp_tokens():
...
+                        external_id=template["id"],
+                    )
+
+                    seen.append(translation)
+
+                # trim any translations we didn't see
+                TemplateTranslation.trim(channel, seen)
+
+            except Exception as e:  # pragma: no cover
+                logger.error("error fetching templates for whatsapp channel: %s" % str(e))
```
Comment (rowanseymour) [inline temba/channels/types/whatsapp/tasks.py]: f-string

**Pull request merged**